cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/table/StructuredTableSpecification.java

```diff
@@ -51,16 +52,26 @@ public final class StructuredTableSpecification {
   private final List<FieldType> fieldTypes;
   private final List<String> primaryKeys;
   private final List<String> indexes;
+  private final boolean needToAddFields;
```

Comment:
You should not rely on this field to determine whether updating columns is needed since it can be different on versions. It will be misleading and confusing for upgrade. The upgrade can happen on any version and this field can become inaccurate. For example,  think about following scenarios:
1. Someone upgrades instance from 6.7 -> 6.8, then this field will be accurate and you can rely on it.
2. Someone upgrades from 6.7 -> 7.0, this is same as scenario 1
3. Someone upgrades from 6.7 -> 6.8, and then 6.8 -> 7.0(assuming no table change), so on second upgrade, this field will be misleading, since no table update is actually needed. This will be worse if later we support deleting, changing primary keys.

Reply:
Some alternatives: 1. If you really want the spec to handle this, introduce a version field here, for old tables, this field will be null, for newly changed tables, this becomes "v1" or "v2", every time a table changes happens, we increment this version. Then by comparing the versions, we can know whether an update is needed

Reply:
2. The `StructuredTableAdmin` will be responsible to detect whether a spec is still valid, and allowed to update. This can be done by comparing the old table's schemas with new spec, if there are new columns, we update the table.

Reply:
Depending on what Terence's suggest, feel approach 2 is better if table registry is removed.

Reply:
Cool I'll let the admin deal with schema detect and upgrade. Also I agree that the registry table should be replaced with a live CacheLoader, since CacheLoader connects db and read the latest schema, the registry table adds more in-directory for sql tables.